sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -0,0 +1,360 @@
+
+	ret = of_reserved_mem_device_init(dev);
+	if (ret) {
+		dev_err(dev, "of_reserved_mem_device_init failed\n");
+		return ret;
+	}
+
+	mem_region = of_parse_phandle(dev->of_node, "memory-region", 1);
+	if (!mem_region) {
+		dev_err(dev, "no 'memory-region' phandle\n");
+		return -ENODEV;
+	}
+
+	ret = of_address_to_resource(mem_region, 0, res);
+	if (ret) {
+		dev_err(dev, "of_address_to_resource failed\n");
+		return ret;
+	}
+
+	adsp->pa_dram = (phys_addr_t)res->start;
+	adsp->dramsize = (u32)resource_size(res);
+	if (((u32)adsp->pa_dram) & DRAM_REMAP_MASK) {
+		dev_err(dev, "adsp memory(0x%x) is not 4K-aligned\n",
+			(u32)adsp->pa_dram);
+		return -EINVAL;
+	}
+
+	if (adsp->dramsize < TOTAL_SIZE_SHARED_DRAM_FROM_TAIL) {
+		dev_err(dev, "adsp memory(0x%x) is not enough for share\n",
+			(u32)adsp->dramsize);
+		return -EINVAL;
+	}
+
+	dev_dbg(dev, "[ADSP] dram pbase=%pa, dramsize=0x%x\n",
+		&adsp->pa_dram, adsp->dramsize);
+
+	/* Parse CFG base */
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!res) {
+		dev_err(dev, "no ADSP-CFG register resource\n");
+		return -ENXIO;
+	}
+	/* remap for DSP register accessing */
+	adsp->va_cfgreg = devm_ioremap_resource(dev, res);
+	if (IS_ERR(adsp->va_cfgreg))
+		return PTR_ERR(adsp->va_cfgreg);
+
+	adsp->pa_cfgreg = (phys_addr_t)res->start;
+	adsp->cfgregsize = (u32)resource_size(res);
+
+	dev_dbg(dev, "[ADSP] cfgreg-vbase=%p, cfgregsize=0x%x\n",
+		adsp->va_cfgreg, adsp->cfgregsize);
+
+	/* Parse SRAM */
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+	if (!res) {
+		dev_err(dev, "no SRAM resource\n");
+		return -ENXIO;
+	}
+
+	adsp->pa_sram = (phys_addr_t)res->start;
+	adsp->srammsize = (u32)resource_size(res);
+	if (adsp->srammsize < TOTAL_SIZE_SHARED_SRAM_FROM_TAIL) {
+		dev_err(dev, "adsp SRAM(0x%x) is not enough for share\n",
+			(u32)adsp->srammsize);
+		return -EINVAL;
+	}
+
+	dev_dbg(dev, "[ADSP] sram pbase=%pa,0x%x\n", &adsp->pa_sram, adsp->srammsize);
+
+	/* Parse MBOX base */
+	for (i = 0; i < DSP_MBOX_NUM; i++) {
+		res = platform_get_resource(pdev, IORESOURCE_MEM, 2 + i);
+		if (!res) {
+			dev_err(dev, "no MBOX resource %d\n", i);
+			return -ENXIO;
+		}
+
+		adsp->pa_mboxreg[i] = (phys_addr_t)res->start;
+		adsp->va_mboxreg[i] = devm_ioremap_resource(dev, res);
+		if (IS_ERR(adsp->va_mboxreg[i]))
+			return PTR_ERR(adsp->va_mboxreg[i]);
+
+		dev_dbg(dev, "[ADSP]MBOX%d pa_mboxreg:%pa, va_mboxreg:%p\n",
+			i, &adsp->pa_mboxreg[i], adsp->va_mboxreg[i]);
+	}
+
+	return ret;
+}
+
+static int adsp_sram_power_on(struct device *dev, bool val)
+{
+	void __iomem *va_dspsysreg;
+
+	va_dspsysreg = devm_ioremap(dev, ADSP_SRAM_POOL_CON, 0x4);
+	if (!va_dspsysreg) {
+		dev_err(dev, "error: failed to ioremap sram pool base 0x%x\n",
+			ADSP_SRAM_POOL_CON);
+		return -ENOMEM;
+	}
+
+	if (val)
+		writel(readl(va_dspsysreg) & ~DSP_SRAM_POOL_PD_MASK, va_dspsysreg);
+	else
+		writel(readl(va_dspsysreg) | DSP_SRAM_POOL_PD_MASK, va_dspsysreg);
+
```

Comment:
One question here (and I think this is also what sparse warning trying to suggest). You are remapping the IO space with devm_ioremap at line 167 but then you are treating the `va_dspsysreg` as a port mapped address. 

Why aren't using normal read write operations to `va_dspsysreg`?

Reply:
@dbaluta I try to modify it but there is compiler error :  
error: invalid operands to binary expression ('void *' and 'int')  va_dspsysreg |= DSP_SRAM_POOL_PD_MASK;
and I reference other codes in linux kernel. e.g. https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/rc32434_wdt.c#L270 . It looks like also use readl/writel after devm_ioremap.  thanks.